uspace/app/bdsh/cmds/modules/ls/ls.c

```diff
@@ -87,14 +95,46 @@ static unsigned int ls_start(ls_job_t *ls)
  *
  * @param de		Directory element.
  */
-static void ls_print(struct dir_elem_t *de)
+static bool ls_print(struct dir_elem_t *de)
 {
-	if (de->s.is_file)
-		printf("%-40s\t%llu\n", de->name, (long long) de->s.size);
-	else if (de->s.is_directory)
-		printf("%-40s\t<dir>\n", de->name);
+	int width = 13;
+
+	if (de->s.is_file) {
+		if (ls.exact_size) {
+			printf("%-40s\t%*llu\n", de->name, width, (long long) de->s.size);
+		} else {
+			cap_spec_t cap;
+
+			cap_from_blocks(de->s.size, 1, &cap);
+			cap_simplify(&cap);
+
+			char *rptr;
+			if (cap_format(&cap, &rptr) == EOK) {
+				char bytes[9], suffix[3];
+				sscanf(rptr, "%s %s", bytes, suffix);
```

Comment:
Using `sscanf()` here is an overkill. It's also unsafe. The static arrays with magic constant length may cause memory corruption, because you didn't inform `sscanf()` how long they are, and the behavior of `cap_format()` can change. Any place with arrays such as these is a secret ticking bomb. Avoid it whenever at all possible. It's best to avoid any variant of `scanf()` as well, if you can help it. They are hard to use correctly.

Instead, you can find the last space with `strrchr()`, overwrite it with `'\0'`, and use the two parts of the buffer as separate strings.